sql: Implement "ESCAPE"

yuzefovich opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26176)
<!-- Reviewable:end -->

knz commented:
Yes this is going in the right direction. However I highly recommend that you split this PR in two; the first to just add the built-in functions, and the 2nd to make the grammar changes and add the parse tests.

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LDmdPQp1KO1QJUwKn6A:b-a5qdu5)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Should I include release notes for these PRs?

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LDmtC0F0Datl8iepts6:b-8p7vcf)*
<!-- Sent from Reviewable.io -->

knz commented:
yes to both. Name the built-in functions in full there.

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LDq487e132KtBwF4F6R:b-qunp7d)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Done. Is it good to merge?

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LDqCmL70v95orkAVvMf:bnk5xtt)*
<!-- Sent from Reviewable.io -->

knz commented:
No, it is not good to merge. We don't merge PRs that add built-in functions unless the functions actually do something. Please continue the work and make the matching work, then we can merge this.

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LDqQ6UM4nh90OKjX6w0:bxlhqq0)*
<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: add built-in functions for ESCAPE clause

Adds built-in functions for LIKE/ILIKE/NOT LIKE/NOT ILIKE/
SIMILAR TO/NOT SIMILAR TO clauses when used with optional
ESCAPE clause.

Incremental change towards implementing #23538.

Release note (sql change): Add built-in functions like_escape,
ilike_escape, not_like_escape, not_ilike_escape, similar_escape,
not_similar_escape to be used when optional ESCAPE clause is present.

knz commented:
Good progress. This is nearly ready to go. Since the change is not a "work in progress" any more you can change the text in the commit message to become more descriptor, for example "Incremental change towards fixing #23538."

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/sem/tree/eval.go, line 4010 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDs-YD87K7GAdVzPm1M:-LDs-YD87K7GAdVzPm1N:b-ka8vg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6f2b013d8906b0b411a862195d69d767f999b74/pkg/sql/sem/tree/eval.go#L4010)):*
> ```Go
> 		case '%':
> 			if escape == `%` {
> 				return func(s string) bool {
> ```

`return nil, ...error...`

here and below

---

*[pkg/sql/sem/tree/eval_test.go, line 355 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDs-ijX5nXHpmjHjiTf:-LDs-ijX5nXHpmjHjiTg:b-4hm6ef) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6f2b013d8906b0b411a862195d69d767f999b74/pkg/sql/sem/tree/eval_test.go#L355)):*
> ```Go
> 
> 		// LIKE with ESCAPE clause
> 		{`'A' LIKE '\A' ESCAPE ''`, `false`},
> ```

In this PR, keep the tests here (in `eval_test.go`) but make them use the built-in functions.

e.g. ```
{`builtin_ilike_escape('a', '\a', '')`, `false`},
```

etc

You can replace all your tests in one go with a suitable search-and-replace with a regular expression.
```


---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LDs0-fg8QU7C6MoH3qX:bjj76ms)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/eval.go, line 4010 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDs-YD87K7GAdVzPm1M:-LDs18bWEsGzQEA2dqkm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6f2b013d8906b0b411a862195d69d767f999b74/pkg/sql/sem/tree/eval.go#L4010)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`return nil, ...error...`

here and below
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval_test.go, line 355 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDs-ijX5nXHpmjHjiTf:-LDs8CWA1SbMjme98XKx:b-bipohu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6f2b013d8906b0b411a862195d69d767f999b74/pkg/sql/sem/tree/eval_test.go#L355)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

In this PR, keep the tests here (in `eval_test.go`) but make them use the built-in functions.

e.g. ```
{`builtin_ilike_escape('a', '\a', '')`, `false`},
```

etc

You can replace all your tests in one go with a suitable search-and-replace with a regular expression.
</blockquote></details>

Done. (I thought it would be a lot hard to come up with such regex than it actually was.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/sem/tree/eval.go, line 4268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBIP13KuHrZJwXGzC:-LDsBIP13KuHrZJwXGzD:boouvb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4268)):*
> ```Go
> // with `\\\\` to be unescaped later by unescapePattern.
> // Escape character must be of length 1 and not `\`.
> func replaceCustomEscape(s, escape string) (string, error) {
> ```

I do not understand what this function does. Please put more words in the explanation and give some examples.

Also inside the function body provide more comments about what is happening with intermediate examples.

---

*[pkg/sql/sem/tree/eval.go, line 4305 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBWwrETVEvc8pFwEI:-LDsBWwrETVEvc8pFwEJ:b-gevv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4305)):*
> ```Go
> 			ret[retIndex+1] = '\\'
> 			ret[retIndex+2] = '\\'
> 			ret[retIndex+3] = '\\'
> ```

I don't see where you guarantee that `ret` is large enough.

---

*[pkg/sql/sem/tree/eval.go, line 4330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBoDiCM7DLxZJaPfP:-LDsBoDiCM7DLxZJaPfQ:b-tpgogk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4330)):*
> ```Go
> 		pattern = strings.Replace(pattern, `%`, `.*`, -1)
> 		pattern = strings.Replace(pattern, `_`, `.`, -1)
> 	}
> ```

`else`

---

*[pkg/sql/sem/tree/eval.go, line 4334 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBpMm2H7VjT5zujWJ:-LDsBpMm2H7VjT5zujWK:b-tpgogk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4334)):*
> ```Go
> 		pattern = replaceUnescaped(pattern, `%`, `.*`, `\\`)
> 		pattern = replaceUnescaped(pattern, `_`, `.`, `\\`)
> 	}
> ```

`else`

---

*[pkg/sql/sem/tree/eval.go, line 4339 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBvUv0Tf0ymuYovRr:-LDsBvUw0PaZUnb5tbtW:b-tpgogk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4339)):*
> ```Go
> 		if k.escape != `%` {
> 			pattern = replaceUnescaped(pattern, `%`, `.*`, k.escape)
> 		}
> ```

`else`

---

*[pkg/sql/sem/tree/eval.go, line 4346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBrj632m1SXi2gH4Y:-LDsBrj632m1SXi2gH4Z:bnxoa8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4346)):*
> ```Go
> 	}
> 
> 	if k.escape != `` && k.escape != `\` {
> ```

merge this `if` with the code that precedes

---

*[pkg/sql/sem/tree/eval.go, line 4366 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsC4D239_FX2ghZYkW:-LDsC4D239_FX2ghZYkX:b-4z62xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4366)):*
> ```Go
> 		// whereas actual escape character is not `\`. Case when pattern ends with an actual
> 		// escape character is handled in replaceCustomEscape.
> 		if pattern, err = unescapePattern(pattern, `\\`); err != nil && k.escape == `\` {
> ```

why do you test `k.escape` in the case where `err != nil`? Is it not too late then?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/eval.go, line 4268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBIP13KuHrZJwXGzC:-LDwvOlg93S0RsrKb2ZL:b-5wzdpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4268)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I do not understand what this function does. Please put more words in the explanation and give some examples.

Also inside the function body provide more comments about what is happening with intermediate examples.
</blockquote></details>

I agree, the description was pretty bad and lacking clarity. I expanded it significantly with examples. I added a lot of informative comments as well, so now this function should be more comprehensible.

---

*[pkg/sql/sem/tree/eval.go, line 4305 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBWwrETVEvc8pFwEI:-LDsIHowAU4q1Xaiu0ct:br7h7h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4305)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't see where you guarantee that `ret` is large enough.
</blockquote></details>

I calculate the exact length of the resulting pattern in the beginning.

I added explanations to show why that particular calculation makes sense. I make an assumption that replaceCustomEscape will be called only after QuoteMeta and with non-empty escape character not equal to '\'.

---

*[pkg/sql/sem/tree/eval.go, line 4330 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBoDiCM7DLxZJaPfP:-LDsG8Z75dKFMfxwKULz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4330)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`else`
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4334 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBpMm2H7VjT5zujWJ:-LDsG9CC0q7U0XLulypG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4334)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`else`
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4339 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBvUv0Tf0ymuYovRr:-LDsGLYG67UteddXTPuM:b-y7l3q7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4339)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`else`
</blockquote></details>

We cannot combine these two if blocks because for all escape characters not equal to '', '\', '_', '%' we want to replace both wildcards in the pattern.

---

*[pkg/sql/sem/tree/eval.go, line 4346 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBrj632m1SXi2gH4Y:-LDsH6L67gBWFgOJQD8U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4346)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

merge this `if` with the code that precedes
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4366 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsC4D239_FX2ghZYkW:-LDsH9HR6s6i9GgHBU93:bc3d43k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4366)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

why do you test `k.escape` in the case where `err != nil`? Is it not too late then?
</blockquote></details>

It is not too late - with a custom escape character, an error will have been already found within replaceCustomEscape method.
For example, with '-' as the escape character on pattern 'abc\\' we do not want to return an error 'pattern ends with escape character' because '\\' is not an escape character in this case.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: 1 of 4 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/eval.go, line 4305 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBWwrETVEvc8pFwEI:-LDwzB2F1panneWa_SdI:b-4h7n1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4305)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I calculate the exact length of the resulting pattern in the beginning.

I added explanations to show why that particular calculation makes sense. I make an assumption that replaceCustomEscape will be called only after QuoteMeta and with non-empty escape character not equal to '\'.
</blockquote></details>

*not equal to backslash.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

knz commented:
@nvanbenschoten: Nathan you've written some of the original code and reviewed Richard's recent improvements, so I'd like that you check this new patch is properly aligned with the previous changes.

In particular I wonder if the logic for `QuoteMeta` and `replaceUnescaped` could be better somehow. This dance with doubly-escaping the escape character to un-escape it right away feels intimidating (to me). I'd take any suggestion you may have.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sem/tree/eval.go, line 4268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBIP13KuHrZJwXGzC:-LE8eePd5hFoNjMiU_55:b-rpb9a3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4268)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I agree, the description was pretty bad and lacking clarity. I expanded it significantly with examples. I added a lot of informative comments as well, so now this function should be more comprehensible.
</blockquote></details>

Yes the new comment is *much* better and kudos for also placing all the examples as unit tests. Thank you!

How did you test all these cases with postgres?

---

*[pkg/sql/sem/tree/eval.go, line 4366 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsC4D239_FX2ghZYkW:-LE8cr-03ocHp0JWxOKd:b-5jdusy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4366)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It is not too late - with a custom escape character, an error will have been already found within replaceCustomEscape method.
For example, with '-' as the escape character on pattern 'abc\\' we do not want to return an error 'pattern ends with escape character' because '\\' is not an escape character in this case.
</blockquote></details>

Then pass `k.escape` as argument to `unescapePattern` (specifically, pass `k.escape == '\'` as a `bool` argument suitably named) and make `unescapePattern` avoid generating an error in that case. It is not the caller's responsibility to handle a non-error case that is part of the `unescapePattern` logic.

---

*[pkg/sql/sem/tree/eval.go, line 2236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8aYUF13YuSCwXkbUa:-LE8aYUGDft5w7vJ4b_p:b56n4eu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L2236)):*
> ```Go
> 
> // MatchLikeEscape matches 'unescaped' with 'pattern' using custom escape character 'escape'.
> // 'escape' must be either empty (which disables escape mechanism) or a single character.
> ```

"which disables *the* escape mechanism"

---

*[pkg/sql/sem/tree/eval.go, line 4269 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8eYdVCZIBLIojAzop:-LE8eYdVCZIBLIojAzoq:bm1kc35) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4269)):*
> ```Go
> 
> // Replaces all custom escape characters in s with `\\` only when they are unescaped. 					(1)
> // E.g. original pattern			after QuoteMeta			after replaceCustomEscape with '@' as escape
> ```

The spacing of the column headers is a little off.

---

*[pkg/sql/sem/tree/eval.go, line 4320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8bX8f4SOEOic2iYLd:-LE8bX8f4SOEOic2iYLe:b79c38) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4320)):*
> ```Go
> 
> 	retLen := sLen + unescapedEscapeCount - escapedEscapeCount + (2 * doubleBackslashesCount)
> 	ret := make([]byte, retLen)
> ```

You're trading the benefit to know the exact length in advance by 3 (!!) linear scans of the string upfront, a mandatory cost even if the escapes do not exist in the pattern.
This is not a good trade-off. 

I would recommend you simply allocate the string with some optimistic capacity (`make([]byte, 0, len(s))`) and then `append` to it below.

@nvanbenschoten what do you think?

---

*[pkg/sql/sem/tree/eval.go, line 4393 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8c1J_9k6zlgiuqw9R:-LE8c1J_9k6zlgiuqw9S:bbijnuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4393)):*
> ```Go
> 		pattern = replaceUnescaped(pattern, `%`, `.*`, `\\`)
> 		pattern = replaceUnescaped(pattern, `_`, `.`, `\\`)
> 	} else if k.escape != `` && k.escape != `\` {
> ```

you've already tested `k.escape == '\'` and `k.escape == ''` above, so you do not need this condition here. But you can add a comment to remind the reader that this will indeed hold here.

---

*[pkg/sql/sem/tree/eval.go, line 4424 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8caA7BzTyn-h-8ZPm:-LE8caA7BzTyn-h-8ZPn:b-kwpslz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4424)):*
> ```Go
> 	// other escaped characters `\A` (now `\\A`).
> 	if k.escape != `` {
> 		// We do not want to return an error when pattern ends with supposed escape character `\`
> ```

"with the supposed escape character"

---

*[pkg/sql/sem/tree/eval.go, line 4425 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8ccmc0W3qUjID2R0G:-LE8ccmd6hwnbevhRyFp:bsr7l64) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4425)):*
> ```Go
> 	if k.escape != `` {
> 		// We do not want to return an error when pattern ends with supposed escape character `\`
> 		// whereas actual escape character is not `\`. Case when pattern ends with an actual
> ```

"whereas the actual escape"
"The case when the pattern ends"

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LE8eTPVD25ZkhfKL4TO:b-o94s9w)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/eval.go, line 4268 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsBIP13KuHrZJwXGzC:-LE9yJ294hHQiNouj5jZ:bb6cnuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4268)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes the new comment is *much* better and kudos for also placing all the examples as unit tests. Thank you!

How did you test all these cases with postgres?
</blockquote></details>

Which "cases" are you referring to? All tests I added to `eval_test.go`? Also, I should put unit tests of `replaceCustomEscape` into `eval_internal_test.go`?

---

*[pkg/sql/sem/tree/eval.go, line 4366 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LDsC4D239_FX2ghZYkW:-LEA1WHL-GuSPiLt-b5i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee1d92ad4262e426e41671cf304b47bec67546e8/pkg/sql/sem/tree/eval.go#L4366)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Then pass `k.escape` as argument to `unescapePattern` (specifically, pass `k.escape == '\'` as a `bool` argument suitably named) and make `unescapePattern` avoid generating an error in that case. It is not the caller's responsibility to handle a non-error case that is part of the `unescapePattern` logic.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 2236 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8aYUF13YuSCwXkbUa:-LE9xfeO6RpOBqHZfHky:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L2236)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"which disables *the* escape mechanism"
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4269 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8eYdVCZIBLIojAzop:-LE9yZocD8jvNaSyVtHN:bee3w5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4269)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The spacing of the column headers is a little off.
</blockquote></details>

Eh, tabs :(

---

*[pkg/sql/sem/tree/eval.go, line 4320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8bX8f4SOEOic2iYLd:-LE9yhF25VRwFj0kdowL:bh3twnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4320)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

You're trading the benefit to know the exact length in advance by 3 (!!) linear scans of the string upfront, a mandatory cost even if the escapes do not exist in the pattern.
This is not a good trade-off. 

I would recommend you simply allocate the string with some optimistic capacity (`make([]byte, 0, len(s))`) and then `append` to it below.

@nvanbenschoten what do you think?
</blockquote></details>

I thought allocating some extra space is a worse trade-off, but I do not have enough knowledge to support my hunch.

I can replace these calculations with optimistic `2*len(s)` (which is exact when `s` contains only double backslashes).

---

*[pkg/sql/sem/tree/eval.go, line 4393 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8c1J_9k6zlgiuqw9R:-LE9wD3Z6z2N2oDLJEWl:bohqzsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4393)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you've already tested `k.escape == '\'` and `k.escape == ''` above, so you do not need this condition here. But you can add a comment to remind the reader that this will indeed hold here.
</blockquote></details>

Agreed, done.

---

*[pkg/sql/sem/tree/eval.go, line 4424 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8caA7BzTyn-h-8ZPm:-LE9wcorDMU28gxNs59J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4424)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"with the supposed escape character"
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4425 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8ccmc0W3qUjID2R0G:-LE9wyu_5M6FZHuve1bc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4425)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"whereas the actual escape"
"The case when the pattern ends"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Reviewed 1 of 4 files at r5, 1 of 3 files at r6, 1 of 2 files at r7.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/sem/builtins/builtins.go, line 470 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEv7hIfCEMkwDoMuGGp:-LEv7hIgC4K1r4wOmfQZ:b-xh4vip) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/builtins/builtins.go#L470)):*
> ```Go
> 	},
> 
> 	"like_escape": {
> ```

nit: Kind of strange that these are right in between a series of network related builtins. I'd move them around somewhere.

---

*[pkg/sql/sem/tree/eval.go, line 4269 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8eYdVCZIBLIojAzop:-LEvBR-53KiTJyJl8-8F:b-upuasd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4269)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Eh, tabs :(
</blockquote></details>

What's up with the `(1)` way off at the end of the line?

---

*[pkg/sql/sem/tree/eval.go, line 4320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8bX8f4SOEOic2iYLd:-LEvD-yH7jd7yulI9DQW:b6vluwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4320)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I thought allocating some extra space is a worse trade-off, but I do not have enough knowledge to support my hunch.

I can replace these calculations with optimistic `2*len(s)` (which is exact when `s` contains only double backslashes).
</blockquote></details>

Yeah, I think allocating the max `2*len(s)` and then truncating at the end is a good approach. Allocating `make([]byte, 0, len(s))` seems strange to me because we'll always have to re-allocate unless nothing is escaped, in which case we don't need to copy at all.

I do think the preparation scans would be worth it if it would allow us to detect the case where we can return `s` unchanged without ever allocating. If that's possible then my vote would go to performing the prep scans but fusing them all into a single scan instead of using `strings.Count`.

---

*[pkg/sql/sem/tree/eval.go, line 2236 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEv80yo9E7UTVCk_-P7:-LEv80yo9E7UTVCk_-P8:b-a860wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L2236)):*
> ```Go
> 
> // MatchLikeEscape matches 'unescaped' with 'pattern' using custom escape character 'escape'.
> // 'escape' must be either empty (which disables the escape mechanism) or a single character.
> ```

A single utf-8 character? A single unicode character? Worth deciding and specifying.

---

*[pkg/sql/sem/tree/eval.go, line 4275 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEvCB_S0KcX0hWpnjUm:-LEvCB_S0KcX0hWpnjUn:btb2361) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L4275)):*
> ```Go
> // When an escape character is escaped, we replace it with its single occurrence.								(2)
> // E.g. original pattern       after QuoteMeta       after replaceCustomEscape with '@' as escape
> //        '@@w@w'         ->      '@@w@w'       ->        '@w||w'
> ```

`s/||/\\/`

---

*[pkg/sql/sem/tree/eval.go, line 4419 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEvAvXW06SgI84LyHk5:-LEvAvXW06SgI84LyHk6:btb2361) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L4419)):*
> ```Go
> 	}
> 
> 	// After QuoteMeta, all '\' were converted to '||'.
> ```

`s/||/\\/`

---

*[pkg/sql/sem/tree/eval.go, line 4449 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEv9Kei8kYF5-69GGPq:-LEv9Kei8kYF5-69GGPr:bwcpdca) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L4449)):*
> ```Go
> 		// We need to provide non-empty escape character to make type checking
> 		// happy, but it won't be used.
> 		pattern = similarEscapeCustomChar(k.s, rune('\\'), false)
> ```

Just pass `0` instead of passing what looks like a valid escape character (`rune('\\')`).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/sem/builtins/builtins.go, line 470 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEv7hIfCEMkwDoMuGGp:-LEyZkqP0veH4erQI9gG:b-8bzq0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/builtins/builtins.go#L470)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Kind of strange that these are right in between a series of network related builtins. I'd move them around somewhere.
</blockquote></details>

I moved them right below regexp related builtins, not sure if there is a more suitable place.

---

*[pkg/sql/sem/tree/eval.go, line 4269 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8eYdVCZIBLIojAzop:-LEyaKwMD1g_mkOQrLwK:bz4it5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4269)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's up with the `(1)` way off at the end of the line?
</blockquote></details>

I use these markers to refer to specific rules in the comments to make the code easier to comprehend.

---

*[pkg/sql/sem/tree/eval.go, line 4320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LE8bX8f4SOEOic2iYLd:-LEyiPoj7tK6hDh1NARL:b-6vz2rv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6deba556d224478f60115bd82ceed008f56afc0b/pkg/sql/sem/tree/eval.go#L4320)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, I think allocating the max `2*len(s)` and then truncating at the end is a good approach. Allocating `make([]byte, 0, len(s))` seems strange to me because we'll always have to re-allocate unless nothing is escaped, in which case we don't need to copy at all.

I do think the preparation scans would be worth it if it would allow us to detect the case where we can return `s` unchanged without ever allocating. If that's possible then my vote would go to performing the prep scans but fusing them all into a single scan instead of using `strings.Count`.
</blockquote></details>

I combined all prep scans into one along with error checking and checking whether `s` remains unchanged.

---

*[pkg/sql/sem/tree/eval.go, line 2236 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEv80yo9E7UTVCk_-P7:-LEy_X9X-vlKCmMAYwe9:bu8v6s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L2236)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

A single utf-8 character? A single unicode character? Worth deciding and specifying.
</blockquote></details>

I believe it should be a single unicode character. We don't need to make any changes in the code to enforce that, right?

---

*[pkg/sql/sem/tree/eval.go, line 4275 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEvCB_S0KcX0hWpnjUm:-LEybaJO9DJsDL4WYYij:b-m16fak) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L4275)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/||/\\/`
</blockquote></details>

Nice catch, thanks.

---

*[pkg/sql/sem/tree/eval.go, line 4419 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEvAvXW06SgI84LyHk5:-LEybeA26_WeMEKo7eAm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L4419)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/||/\\/`
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4449 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LEv9Kei8kYF5-69GGPq:-LEybmZZ-cCXWbf-X8wB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/453ff38a0762e253c18fb3a5007cfc1c691e5e8e/pkg/sql/sem/tree/eval.go#L4449)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Just pass `0` instead of passing what looks like a valid escape character (`rune('\\')`).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176)*
<!-- Sent from Reviewable.io -->

knz commented:
Regarding the escape character it might be a plan to use a `rune` instead of a string.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> this looks really good now. Thanks for the work on multi-byte characters.

---

Reviewed 4 of 5 files at r9.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/sem/tree/eval.go, line 2286 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LFOnNdIDwIDLPYpHs1P:-LFOnNdJ-sGGWrmlgOAz:b7fzeab) ([raw file](https://github.com/cockroachdb/cockroach/blob/5560cbcef5ca2c689d3992b774056aa1dd206049/pkg/sql/sem/tree/eval.go#L2286)):*
> ```Go
> func matchLike(ctx *EvalContext, left, right Datum, caseInsensitive bool) (Datum, error) {
> 	pattern := string(MustBeDString(right))
> 	like, err := optimizedLikeFunc(pattern, caseInsensitive, rune('\\'))
> ```

nit: no need for cast.

---

*[pkg/sql/sem/tree/eval.go, line 4539 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LFOoD2a1JmBadevJGqR:-LFOoD2a1JmBadevJGqS:bea92q6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5560cbcef5ca2c689d3992b774056aa1dd206049/pkg/sql/sem/tree/eval.go#L4539)):*
> ```Go
> 	if k.escape == 0 {
> 		// k.escape is empty which disables the escape mechanism.
> 		pattern = similarEscapeCustomChar(k.s, k.escape, false)
> ```

`pattern := similarEscapeCustomChar(k.s, k.escape, k.escape == 0)`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LFOoS_4Bbbo0_ks5c6d:bx3ejf8)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for all the comments.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/sem/tree/eval.go, line 2286 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LFOnNdIDwIDLPYpHs1P:-LFP0lfu-ybsvRHLuMeC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5560cbcef5ca2c689d3992b774056aa1dd206049/pkg/sql/sem/tree/eval.go#L2286)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: no need for cast.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 4539 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-LFOoD2a1JmBadevJGqR:-LFP19Yy1miMJdag8_bU:b-m16fak) ([raw file](https://github.com/cockroachdb/cockroach/blob/5560cbcef5ca2c689d3992b774056aa1dd206049/pkg/sql/sem/tree/eval.go#L4539)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`pattern := similarEscapeCustomChar(k.s, k.escape, k.escape == 0)`
</blockquote></details>

Nice catch, thanks.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26176#-:-LFP2FZjEGlQziwntbdW:bhpjtkz)*
<!-- Sent from Reviewable.io -->

yuzefovich commented:
bors r+
